river/Seat.zig

```diff
@@ -342,17 +342,49 @@ pub fn handleViewUnmap(self: *Self, view: *View) void {
     if (self.focused == .view and self.focused.view == view) self.focus(null);
 }
 
-/// Handle any user-defined mapping for the passed keysym and modifiers
+/// Handle any user-defined mapping for passed keycode, modifiers and keyboard state
 /// Returns true if the key was handled
 pub fn handleMapping(
     self: *Self,
-    keysym: xkb.Keysym,
+    keycode: xkb.Keycode,
     modifiers: wlr.Keyboard.ModifierMask,
     released: bool,
+    xkb_state: *xkb.State,
 ) bool {
     const modes = &server.config.modes;
+    // TODO:
+    // There are two forms how we can handle a key combination:
+    //      1. Raw keysyms and modifiers as if modifiers didn't change keysyms.
+    //         E.g. pressing `Super+Shift 1` does not translate to `Super Exclam`.
+    //      2. Keysyms and modifiers as translated by xkb.
+    //         Modifiers used to translate the key are consumed.
+    //         E.g. pressing `Super+Shift 1` translates to `Super Exclam`.
+    //
+    // Now we introduce mappings with pinned layouts. These can also be handled in
+    // the above ways.
+    //
+    // This means a user could define multiple mappings that are all triggered by
+    // the exact same key combination.
+    //
+    // Example:
+    //
+    // XKB_DEFAULT_LAYOUT=us,fr river
+    // # Following mappings would all be triggered by the combination `Super+Shift 1`:
+    // map normal Super+Shift 1 ...
+    // map normal Super Exclam ... #Shift is "consumed" to produce Exclam
+    // # In French AZERTY, numbers are on the second level
+    // map -layout 1 normal Super+Shift Ampersand ...
+    // map -layout 1 normal Super 1 ... #Shift is consumed to produce 1
+    //
+    // The question now is: Do we want to have a defined behaviour which one of these
+    // mappings would be triggered in such a case?
+    // If so, we would need to loop over the mappings-list multiple times. Up to 4
+    // times if we want to set a clear priority-list.
+
+    // For now, do everything in one wipe, which means any of these mappings could be
+    // the one that is triggered
```

Comment:
Hmm, if there's an easy way to detect and forbid conflicting mappings as in the example my first choice would be to forbid this entirely. If not, I think the simplest and least surprising thing we can do would be to activate *both* mappings. That is, instead of returning early from the for loop below we would continue in case there is another match. I'd be fine with leaving the ordering between the mappings undefined for simplicity. How's that sound?

Reply:
Sounds good. Implemented it that way.. However, only one mapping can be repeated. We could make `seat.repeating_mapping` a linked list if we cared about this edge case...

Reply:
Ah yeah makes sense :)